onnxruntime/core/providers/nnapi/nnapi_builtin/builders/op_builder.cc

```diff
@@ -1859,8 +1878,16 @@ Status ConcatOpBuilder::AddToModelBuilderImpl(ModelBuilder& model_builder, const
   const auto& input0 = inputs[0].node_arg.Name();
   const auto node_input_size = inputs.size();
 
-  // First if the inputs are uint8, we need verify all the inputs have same scale and zero points
-  if (operand_types.at(input0).type == android::nn::wrapper::Type::TENSOR_QUANT8_ASYMM) {
+  bool is_quant_op = IsQuantizedOp(node_unit);
+
+  // First if the inputs are uint8 and this is not a quantized Concat,
+  // we need verify all the inputs have same scale and zero points
+  // ONNX allows Concat (not QlinearConcat, not QDQ concat) run directly on uint8 without scales and zps,
+  // However, for NNAPI, all int8 inputs need scale and zero points cannot be 0
+  // so we need to use the scales and zps from the NNAPI input directly, there is no easy way to get the input
+  // scales and zps in OpSupportChecker, so we need to verify here
+  // Also we have to assume the output scale and zp are the same as input 0
+  if (operand_types.at(input0).type == android::nn::wrapper::Type::TENSOR_QUANT8_ASYMM && !is_quant_op) {
```

Comment:
Can we check this in ConcatOpSupportChecker::HasSupportedInputOutputsImpl? Looks like we do the check for zp/scale being consistent for a quantized op there.

What's the reason for only checking if input is uint8? Is int8 not possible or that is checked somewhere else?

Reply:
This check is for Concat op, (not the QlinearConcat, not the QDQ Concat) with straight uint8 inputs without scale and zp, there is no easy way to get the scales and zps without construct the NNAPI model, so it's very hard to do in the ConcatOpSupportChecker